lib/cpp/src/thrift/TUuid.h

```diff
@@ -0,0 +1,148 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#ifndef _THRIFT_TUUID_H_
+#define _THRIFT_TUUID_H_ 1
+
+#include <thrift/Thrift.h>
+
+#include <algorithm>
+
+namespace apache {
+namespace thrift {
+
+/**
+ * Thrift wrapper class for a UUID type.
+ *
+ * The UUID is stored as a 16 byte buffer.
+ * This class stores the UUID in network order when assigned from a string.
+ */
+class TUuid {
+public:
+  typedef uint8_t value_type;
+  typedef uint8_t* iterator;
+  typedef uint8_t const* const_iterator;
+  typedef std::size_t size_type;
+  typedef std::ptrdiff_t difference_type;
+
+  TUuid() = default;
+  TUuid(const TUuid& other) = default;
+  TUuid(TUuid&& other) = default;
+  TUuid& operator=(const TUuid&) = default;
+  TUuid& operator=(TUuid&&) = default;
+  ~TUuid() = default;
+
+  /**
+   * Construct the object from the specified string.
+   *
+   * Supported string formats are:
+   *   - "hhhhhhhh-hhhh-hhhh-hhhh-hhhhhhhhhhhh"
+   *   - "{hhhhhhhh-hhhh-hhhh-hhhh-hhhhhhhhhhhh}"
+   *   - "hhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh"
+   *   - "{hhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh}"
+   *
+   * If the string is invalid, the object will be set to a
+   * nil (empty) UUID.
+   */
+  explicit TUuid(const std::string& str) noexcept;
+
+  /**
+   * Copy assignment from a UUID string.
+   *
+   * This function will throw an exception if the string is not
+   * a valid UUID.
+   */
+  TUuid& operator=(const std::string& str) noexcept(false);
```

Comment:
Any comment on this function? 
I am still very much in two minds if this function should ever exist.
Perhaps not since there is an `explicit` constructor taking a string.
I am concerned about the porting issue for people that already used a string like object (byte[16]) but wants to upgrade to using the strong UUID.